server/paymentProviders/stripe/virtual-cards.js

```diff
@@ -51,3 +58,173 @@ export const assignCardToCollective = async (cardNumber, expireDate, cvv, collec
 
   return await models.VirtualCard.create(cardData);
 };
+
+export const processTransaction = async (stripeEvent, stripeSignature) => {
+  const stripeTransaction = stripeEvent.data.object;
+
+  const virtualCard = await models.VirtualCard.findOne({
+    where: {
+      id: stripeTransaction.card,
+    },
+    include: [
+      { association: 'collective', required: true },
+      { association: 'host', required: true },
+      { association: 'user' },
+    ],
+  });
+
+  if (!virtualCard) {
+    logger.error('stripe/webhook: could not find VirtualCard', { body: stripeTransaction });
+    return;
+  }
+
+  const host = virtualCard.host;
+  const collective = virtualCard.collective;
+  const [connectedAccount] = await host.getConnectedAccounts({ where: { service: 'stripe' } });
+
+  if (!connectedAccount) {
+    logger.error('stripe/webhook: host is not connected to Stripe', { body: stripeTransaction });
+    return;
+  }
+
+  const stripe = Stripe(connectedAccount.token);
+
+  try {
+    stripe.webhooks.constructEvent(stripeEvent, stripeSignature, connectedAccount.data.stripeEndpointSecret);
+  } catch {
+    logger.error('stripe/webhook: source of event not recognized', { body: stripeTransaction });
+    return;
+  }
+
+  const amount = stripeTransaction.amount;
+
+  if (amount === 0) {
+    return;
+  }
+
+  const existingExpense = await models.Expense.findOne({
+    where: {
+      VirtualCardId: virtualCard.id,
+      data: { token: stripeTransaction.id },
+    },
+  });
+  if (existingExpense) {
+    logger.warn(`Virtual Card charge already reconciled, ignoring it: ${stripeTransaction.id}`);
+    return;
+  }
+
+  const UserId = virtualCard.UserId;
+  const isRefund = amount < 0;
+
+  // If it is refund, we'll check if the transaction was already created because there are no expenses created for refunds.
+  if (isRefund) {
+    const existingTransaction = await models.Transaction.findOne({
+      where: {
+        CollectiveId: collective.id,
+        data: { token: stripeTransaction.id },
+      },
+    });
+    if (existingTransaction) {
+      logger.warn(`Virtual Card refund already reconciled, ignoring it: ${stripeTransaction.id}`);
+      return;
+    }
+  }
+
+  let expense;
+  try {
+    const slug = toString(stripeTransaction['merchant_data']['network_id']).toLowerCase();
+    const [vendor] = await models.Collective.findOrCreate({
```

Comment:
Why would this be returned in an array?

Reply:
`findOrCreate` return an instance with the object and a boolean. This is the javascript syntax equivalent of :
```
const sequelizeResult = await models.Collective.findOrCreate(...);
const vendor = sequelizeResult[0];
```
See https://sequelize.org/master/manual/model-querying-finders.html

Reply:
Right.